### DISCUSSION TITLE
Deploying Factory in the Account abstraction multisig tutorial

### DISCUSSION THREAD

**Discussion by luisvid** [zkSync-Community-Hub/zksync-developers#204] [Deployments] [answered]
### Environment

Testnet

### zkSolc Version

v1.3.17

### zksync-web3 Version

^0.16.0

### Hardhat.config.ts

```
import { HardhatUserConfig } from "hardhat/config";
import "@matterlabs/hardhat-zksync-deploy";
import "@matterlabs/hardhat-zksync-solc";

import "@matterlabs/hardhat-zksync-verify";

const config: HardhatUserConfig = {
  zksolc: {
    version: "latest", // Uses latest available in https://github.com/matter-labs/zksolc-bin/
    settings: {
      isSystem: true, // make sure to include this line
    },
  },
  defaultNetwork: "zkSyncTestnet",

  networks: {
    zkSyncTestnet: {
      url: "https://testnet.era.zksync.dev",
      ethNetwork: "https://goerli.infura.io/v3/<my_api_key>", // Can also be the RPC URL of the network (e.g. `https://goerli.infura.io/v3/<API_KEY>`)
      zksync: true,
    },
  },
  solidity: {
    version: "0.8.17",
  },
};

export default config;

```

### Deployment Script (WITHOUT PRIVATE KEY)

```
import { utils, Wallet } from "zksync-web3";
import * as ethers from "ethers";
import { HardhatRuntimeEnvironment } from "hardhat/types";
import { Deployer } from "@matterlabs/hardhat-zksync-deploy";

export default async function (hre: HardhatRuntimeEnvironment) {
  // Private key of the account used to deploy
  // const wallet = new Wallet("<WALLET_PRIVATE_KEY>");
  const wallet = new Wallet(
    ""
  );
  const deployer = new Deployer(hre, wallet);
  const factoryArtifact = await deployer.loadArtifact("AAFactory");
  const aaArtifact = await deployer.loadArtifact("TwoUserMultisig");

  // Check that the artifacts were loaded correctly
  if (!factoryArtifact) {
    throw new Error("Failed to load factoryArtifact");
  }
  if (!aaArtifact || !aaArtifact.bytecode) {
    throw new Error("Failed to load aaArtifact or aaArtifact.bytecode");
  }

  // Calculate the bytecode hash
  const bytecodeHash = utils.hashBytecode(aaArtifact.bytecode);

  // Check that the bytecode hash was calculated correctly
  if (!bytecodeHash) {
    throw new Error("Failed to calculate bytecodeHash");
  }

  const factory = await deployer.deploy(
    factoryArtifact,
    [utils.hashBytecode(aaArtifact.bytecode)],
    undefined,
    [aaArtifact.bytecode]
  );

  console.log(`AA factory address: ${factory.address}`);
}

```

### Contract Code

```
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

import "@matterlabs/zksync-contracts/l2/system-contracts/Constants.sol";
import "@matterlabs/zksync-contracts/l2/system-contracts/libraries/SystemContractsCaller.sol";

contract AAFactory {
    bytes32 public aaBytecodeHash;

    constructor(bytes32 _aaBytecodeHash) {
        aaBytecodeHash = _aaBytecodeHash;
    }

    function deployAccount(
        bytes32 salt,
        address owner1,
        address owner2
    ) external returns (address accountAddress) {
        (bool success, bytes memory returnData) = SystemContractsCaller
            .systemCallWithReturndata(
                uint32(gasleft()),
                address(DEPLOYER_SYSTEM_CONTRACT),
                uint128(0),
                abi.encodeCall(
                    DEPLOYER_SYSTEM_CONTRACT.create2Account,
                    (
                        salt,
                        aaBytecodeHash,
                        abi.encode(owner1, owner2),
                        IContractDeployer.AccountAbstractionVersion.Version1
                    )
                )
            );
        require(success, "Deployment failed");

        (accountAddress) = abi.decode(returnData, (address));
    }
}

```

### Does this work on other EVMs? (If yes, please list at least 1 of them)

I didn't try it

### Description of What Your Contract Does

This is the contract factory that deploys the multisig account accounts from the Tutorial Account abstraction multisig https://era.zksync.io/docs/dev/tutorials/custom-aa-tutorial.html

Following all the steps and using the code provided in the tutorial, when I want to deploy the Factory with this line:

`yarn hardhat deploy-zksync --script deploy-factory.ts`

I get the following error

> An unexpected error occurred:
> 
> <ref *1> Error: cannot estimate gas; transaction may fail or may require manual gas limit [ See: https://links.ethers.org/v5-errors-UNPREDICTABLE_GAS_LIMIT ] (error={}, tx={"customData":{"factoryDeps":[null,"0x00020000000000020001000000010355000000000301001900000060033002700000003a0030019d0000003a033001970000000102200190000000940000c13d0000008002000039000000400020043f000000040430008c000000c50000413d000000000401043b000000e0044002700000003d0540009c000000be0000613d0000003e0440009c000000c50000c13d0000000004000416000000000404004b000000c50000c13d000000040330008a000000600330008c000000c50000413d0000002403100370000000000303043b000000400430009c000000c50000213d0000004404100370000000000404043b000000400540009c000000c50000213d0000000005000414000000000600041a000000a00030043f000000c00040043f0000004007000039000000800070043f0000004108000041000001000080043f0000000401100370000000000101043b000001040010043f000001240060043f000001440020043f000001840070043f000001a40030043f000001c40040043f000001e40000043f0000000101000039000001640010043f000000e401000039000000e00010043f0000020001000039000000400010043f000000c001500210000000420110019700000043011001c70000800602000039000000000300001900000000040000190000000005000019000000000600001900e400df0000040f000000000301001900000060033002700000003a053001970000003f035000390000003b06300197000000400400043d0000000003460019000000000663004b00000000060000190000000106004039000000440730009c000000c70000213d0000000106600190000000c70000c13d000000400030043f00000000035404360000001f0650003900000005066002720000005e0000613d0000000007000031000000010770036700000000080000190000000509800210000000000a930019000000000997034f000000000909043b00000000009a04350000000108800039000000000968004b000000560000413d000000000600004b000000600000613d0000001f0650018f00000005055002720000006c0000613d000000000700001900000005087002100000000009830019000000000881034f000000000808043b00000000008904350000000107700039000000000857004b000000640000413d000000000706004b0000007b0000613d0000000505500210000000000151034f00000000055300190000000306600210000000000705043300000000076701cf000000000767022f000000000101043b0000010006600089000000000161022f00000000016101cf000000000171019f00000000001504350000000101200190000000cd0000613d00000000010404330000004802000041000000200410008c000000000400001900000000040240190000004801100197000000000501004b000000000200a019000000480110009c000000000204c019000000000102004b000000c50000c13d0000000001030433000000400210009c000000c50000213d000000400200043d00000000001204350000003a010000410000003a0320009c0000000002018019000000400120021000000049011001c7000000e50001042e0000000002000416000000000202004b000000c50000c13d0000009f023000390000003b02200197000000400020043f0000001f0230018f0000000504300272000000a60000613d00000000050000190000000506500210000000000761034f000000000707043b000000800660003900000000007604350000000105500039000000000645004b0000009e0000413d000000000502004b000000b50000613d0000000504400210000000000141034f00000003022002100000008004400039000000000504043300000000052501cf000000000525022f000000000101043b0000010002200089000000000121022f00000000012101cf000000000151019f0000000000140435000000200130008c000000c50000413d000000800100043d000000000010041b0000002001000039000001000010044300000120000004430000003c01000041000000e50001042e0000000001000416000000000101004b000000c50000c13d000000000100041a000000800010043f0000003f01000041000000e50001042e0000000001000019000000e6000104300000004a0100004100000000001004350000004101000039000000040010043f0000004b01000041000000e600010430000000400100043d000000440210003900000045030000410000000000320435000000240210003900000011030000390000000000320435000000460200004100000000002104350000000402100039000000200300003900000000003204350000003a020000410000003a0310009c0000000001028019000000400110021000000047011001c7000000e600010430000000e2002104210000000102000039000000000001042d0000000002000019000000000001042d000000e400000432000000e50001042e000000e600010430000000000000000000000000000000000000000000000000000000000000000000000000ffffffff00000000000000000000000000000000000000000000000000000001ffffffe0000000020000000000000000000000000000004000000100000000000000000000000000000000000000000000000000000000000000000000000000c795c9130000000000000000000000000000000000000000000000000000000076fb8b650000000000000000000000000000000000000020000000800000000000000000000000000000000000000000ffffffffffffffffffffffffffffffffffffffff5d3827000000000000000000000000000000000000000000000000000000000000000000ffffffff00000000000000000000000000000000000000000000000001000000000000000000000000000000000000e4000001000000000000000000000000000000000000000000000000000000000000000000ffffffffffffffff4465706c6f796d656e74206661696c656400000000000000000000000000000008c379a0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000064000000000000000000000000800000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000200000000000000000000000004e487b7100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002400000000000000000000000002b22facd64b01da8b7239697383a7911d84bcbbdb30e8c2bf7b3b1270991701"],"gasPerPubdata":50000},"data":"0x9c4d535b00000000000000000000000000000000000000000000000000000000000000000100004d8e5c7f4baf6a9d706fa339fcd50a7f641a0b7ebc0a3b50331567b0c500000000000000000000000000000000000000000000000000000000000000600000000000000000000000000000000000000000000000000000000000000020010006338b8a07fc5adf898f8ffde44cfbeca02ddc1f3556889a9dc87c055bc7","to":{},"type":113,"from":"0x1487f4304C8663683BEEf72B32bc8845F5Fb8941","nonce":{},"gasLimit":{},"chainId":{}}, code=UNPREDICTABLE_GAS_LIMIT, version=abstract-signer/5.7.0)
>     at Logger.makeError (/Users/luisvidela/Development/Academy/zkSyncEra/custom-aa/node_modules/@ethersproject/logger/src.ts/index.ts:269:28)
>     at Logger.throwError (/Users/luisvidela/Development/Academy/zkSyncEra/custom-aa/node_modules/@ethersproject/logger/src.ts/index.ts:281:20)
>     at /Users/luisvidela/Development/Academy/zkSyncEra/custom-aa/node_modules/@ethersproject/abstract-signer/src.ts/index.ts:301:31
>     at processTicksAndRejections (node:internal/process/task_queues:95:5)
>     at async Promise.all (index 6) {
>   reason: 'cannot estimate gas; transaction may fail or may require manual gas limit',
>   code: 'UNPREDICTABLE_GAS_LIMIT',
>   error: TypeError: Cannot read properties of undefined (reading 'toHexString')
>       at isHexable (/Users/luisvidela/Development/Academy/zkSyncEra/custom-aa/node_modules/@ethersproject/bytes/src.ts/index.ts:58:21)
>       at Object.arrayify (/Users/luisvidela/Development/Academy/zkSyncEra/custom-aa/node_modules/@ethersproject/bytes/src.ts/index.ts:115:9)
>       at /Users/luisvidela/Development/Academy/zkSyncEra/custom-aa/node_modules/zksync-web3/build/src/provider.js:343:39
>       at Array.map (<anonymous>)
>       at hexlifyTransaction (/Users/luisvidela/Development/Academy/zkSyncEra/custom-aa/node_modules/zksync-web3/build/src/provider.js:340:80)
>       at Provider.JsonRpcProvider.prepareRequest (/Users/luisvidela/Development/Academy/zkSyncEra/custom-aa/node_modules/@ethersproject/providers/src.ts/json-rpc-provider.ts:599:47)
>       at Provider.<anonymous> (/Users/luisvidela/Development/Academy/zkSyncEra/custom-aa/node_modules/@ethersproject/providers/src.ts/json-rpc-provider.ts:634:27)
>       at step (/Users/luisvidela/Development/Academy/zkSyncEra/custom-aa/node_modules/@ethersproject/providers/lib/json-rpc-provider.js:48:23)
>       at Object.next (/Users/luisvidela/Development/Academy/zkSyncEra/custom-aa/node_modules/@ethersproject/providers/lib/json-rpc-provider.js:29:53)
>       at /Users/luisvidela/Development/Academy/zkSyncEra/custom-aa/node_modules/@ethersproject/providers/lib/json-rpc-provider.js:23:71,
>   tx: {
>     customData: { factoryDeps: [Array], gasPerPubdata: 50000 },
>     data: '0x9c4d535b00000000000000000000000000000000000000000000000000000000000000000100004d8e5c7f4baf6a9d706fa339fcd50a7f641a0b7ebc0a3b50331567b0c500000000000000000000000000000000000000000000000000000000000000600000000000000000000000000000000000000000000000000000000000000020010006338b8a07fc5adf898f8ffde44cfbeca02ddc1f3556889a9dc87c055bc7',
>     to: Promise { '0x0000000000000000000000000000000000008006' },
>     type: 113,
>     from: '0x1487f4304C8663683BEEf72B32bc8845F5Fb8941',
>     nonce: Promise { 2 },
>     gasLimit: Promise { <rejected> [Circular *1] },
>     chainId: Promise { 280 }
>   }
> }
> error Command failed with exit code 1.
> 

To solve it I want to manually pass the gas limit when deploying, as follows in the Factory deploy

```
  const factory = await deployer.deploy(
    factoryArtifact,
    [utils.hashBytecode(aaArtifact.bytecode)],
    { gasLimit: '5000000' },
    [aaArtifact.bytecode]
  );
```

and when I do I get the following error

> TypeError: Cannot read properties of undefined (reading 'toHexString')
>     at isHexable (/Users/luisvidela/Development/Academy/zkSyncEra/custom-aa/node_modules/@ethersproject/bytes/src.ts/index.ts:58:21)
>     at Object.arrayify (/Users/luisvidela/Development/Academy/zkSyncEra/custom-aa/node_modules/@ethersproject/bytes/src.ts/index.ts:115:9)
>     at hashBytecode (/Users/luisvidela/Development/Academy/zkSyncEra/custom-aa/node_modules/zksync-web3/build/src/utils.js:205:51)
>     at /Users/luisvidela/Development/Academy/zkSyncEra/custom-aa/node_modules/zksync-web3/build/src/signer.js:51:194
>     at Array.map (<anonymous>)
>     at Function.getSignInput (/Users/luisvidela/Development/Academy/zkSyncEra/custom-aa/node_modules/zksync-web3/build/src/signer.js:51:156)
>     at EIP712Signer.sign (/Users/luisvidela/Development/Academy/zkSyncEra/custom-aa/node_modules/zksync-web3/build/src/signer.js:56:111)
>     at processTicksAndRejections (node:internal/process/task_queues:95:5)
>     at async Wallet.signTransaction (/Users/luisvidela/Development/Academy/zkSyncEra/custom-aa/node_modules/zksync-web3/build/src/wallet.js:92:54)
> error Command failed with exit code 1.

Why does the original error occur and what is the correct way to pass the gasLimit to deployer.deploy()?
Thanks

**Comment by AlbionaHoti** [comment]
Hi @luisvid, thanks for opening this issue! Is this still occurring? If you've followed the tutorial steps it should be working fine because lately, we've been testing everything with our sepolia migration, but to make sure, I'll again try and get back to you!

**Reply by luisvid** [reply]
Hi @AlbionaHoti , how are you?

Honestly, that was almost 2 months ago and after not getting a response, I gave up. It's quite frustrating to follow a tutorial and have it not work. Now that you mention you've tried it and it works, I'm willing to give it another shot.
Thanks!!

**Reply by AlbionaHoti** [reply, marked as answer]
Hi @luisvid, sorry for the frustration caused by the lengthy tutorial. We are working on making the tutorial shorter and easier to understand. Please come back if you find any other issues following the Account abstraction one. We've updated the docs in a new structure. I hope you're going to find it helpful.

**Comment by varnerrobertv** [comment]
The original error occurs because the gas limit cannot be estimated automatically, and it may require a manual gas limit to be set. To pass the gas limit to deployer.deploy(), you should specify it as an option object within the third argument of the function call. However, the correct property name for the gas limit is gasPrice, not gasLimit. So, you should use { gasPrice: '5000000' } instead of { gasLimit: '5000000' }. This should resolve the error you encountered.

**Comment by quanteauphil** [comment]
The original error occurs because the gas limit cannot be estimated automatically, and the transaction may fail due to unpredictable gas usage. To solve this, manually pass the gas limit when deploying using deployer.deploy() by providing it as an option object { gasLimit: '5000000' }. However, ensure that the gas limit is provided as a BigNumber or a string representing a hexadecimal value, not a number. Additionally, ensure that all promises for the to, nonce, and chainId parameters are resolved before passing them to the deploy() function to avoid the "TypeError: Cannot read properties of undefined" error.

**Comment by mcgeesusan** [comment]
The original error occurs due to an issue with estimating gas for the transaction, possibly due to the zkSync provider. To manually pass gas limit to deployer.deploy(), ensure the gasLimit is provided as a BigNumber or a number, not a string. Use a value appropriate for your deployment scenario.

**Comment by matthewredmanv** [comment]
The original error occurs due to the inability to estimate gas, which may lead to transaction failure. The correct way to pass gasLimit to deployer.deploy() is by providing it as an option object: { gasLimit: '5000000' }.
